docs/animation/asm.md

```diff
@@ -0,0 +1,307 @@
+that callback is in different places depending on what is being animated, although they are always client side, so using them to control server-side things will require some networking code.
+
+
+Code API
+----------
+
+!!! warning
+
+    The ASM code API can only be used _client side_. When storing ASMs in code, use the side-irrelevant `IAnimationStateMachine` interface.
+   
+
+ASMs can be loaded by calling `ModelLoaderRegistry.loadASM`. It takes two parameters, first being a `ResourceLocation` denoting 
+where the ASM is stored, and second an ImmutableMap of load-time defined parameters.
+
+An example:
+```java
+@Nullable
+private final IAnimationStateMachine asm;
+private final VariableValue cycle = new VariableValue(4);
+
+public Spin() {
+     asm = proxy.loadASM(new ResourceLocation(MODID, "asms/block/rotatest.json"), ImmutableMap.of("cycle_length", cycle));
+}
+```
+
+Here, an asm is loaded (from a sidedproxy to avoid crashing on server) with one extra parameter, named `cycle_length`. This parameter is of the type `VariableValue`, so we can
+set it from within our code.
+
+Using an ASM instance, you can get the current state with `.currentState()` and transition to another state with `.transition(nextState)`
+ 
+`VariableValue` parameters can have their value set by calling `.setValue`, but oddly you can not read this value back. There is no need
+to inform the ASM of this change, it happens automatically. All of the other clips can be used from code too if you so desire, but I will not
+go into their API here as they are mainly designed to be used in the JSON files.
+
+File Format
+-------------
+
+The ASMs have a slightly strange and confusing file format, but here's how it works.
+
+First, a simple example:
+```json
+{
+  "parameters": {
+    "anim_cycle": ["/", "#cycle_length"]
+  },
+  "clips": {
+    "default": ["apply", "forgedebugmodelanimation:block/rotatest@default", "#anim_cycle" ]
+  },
+  "states": [
+    "default"
+  ],
+  "transitions": {},
+  "start_state": "default"
+}
+```
+
+As stated above, the files have parameters, clips, states and transitions, as well as the starting state.
+All of these tags are required, even if they are empty.
+
+### Parameters
+
+```javascript
+{
+    "name": <parameter_definition>
+}
+```
+
+Different types of parameters have different formats for `<parameter_definition>`, and the simple ones are:
+
+- `IdentityValue`: the string `#identity`,
+- `ParameterValue`: the parameter to reference, prefixed with `#`, e.g. `#my_awesome_parameter`
+- `ConstValue`: a number to use as the constant to return
+
+Other kinds are a bit more complicated
+
+#### `SimpleExprValue`
```

Comment:
In my opinion, it makes more sense to leave the names as they are in the code since being able to look up the implementation may be useful for people.

Reply:
It's okay to mention the implementation name (in parentheses maybe), but the primary focus should be on the high level concepts and this for example represents mathematical expressions for useful calculations, while `SimpleExprValue` doesn't carry much information.

Besides, most people reading these docs won't ever see any of the parameter implementation classes except for `VariableValue`. The point of documenting this is to not have to look at the implementation.

Reply:
Ok, I see what you mean.